File: supplementary-materials/langchain-vs-springai/backend/python/EssayPreprocessor.py
import re

class EssayPreprocessor:
    def __init__(self):
        self.whitespace_re = re.compile(r'\s+')
        self.punctuation_re = re.compile(r'\s([.,;:!?])')
        self.question_words = {
            'discuss', 'extent', 'advantages', 'disadvantages',
            'opinion', 'view', 'agree', 'disagree'
        }

    def clean_essay(self, essay: str) -> str:
        if not essay:
            return ""
        cleaned = self.whitespace_re.sub(' ', essay)
        return self.punctuation_re.sub(r'\1', cleaned).strip()

    def count_words(self, text: str) -> int:
        return len(text.split()) if text else 0

    def extract_main_topic(self, question: str) -> str:
        if not question:
            return "general"
        
        lower_q = question.lower()
        for word in self.question_words:
            lower_q = lower_q.replace(word, '')
        
        sentences = re.split(r'[.,]', lower_q)
        first_sentence = sentences[0].strip() if sentences else ""
        return first_sentence or "general"

File: supplementary-materials/langchain-vs-springai/backend/java/src/main/java/org/nc/IELTSChecker/services/EssayPreprocessor.java
package org.nc.IELTSChecker.services;

import org.springframework.stereotype.Service;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Pattern;

@Service
public class EssayPreprocessor {

    private static final Pattern WHITESPACE_PATTERN = Pattern.compile("\\s+");
    private static final Pattern PUNCTUATION_PATTERN = Pattern.compile("\\s([.,;:!?])");
    private static final List<String> QUESTION_WORDS = Arrays.asList(
            "discuss", "to what extent", "advantages", "disadvantages",
            "opinion", "view", "agree", "disagree"
    );

    /**
     * Cleans the essay by removing extra whitespace and punctuation.
     * @param essay
     * @return
     */
    public String cleanEssay(String essay) {
        if (essay == null) return "";

        String cleaned = WHITESPACE_PATTERN.matcher(essay).replaceAll(" ");
        cleaned = PUNCTUATION_PATTERN.matcher(cleaned).replaceAll("$1");
        return cleaned.trim();
    }

    /**
     * Counts the number of words in the essay.
     * @param text
     * @return
     */
    public int countWords(String text) {
        if (text == null || text.isEmpty()) return 0;
        return text.split("\\s+").length;
    }

    /**
     * Extracts the main topic from the question.
     * @param question
     * @return
     */
    public String extractMainTopic(String question) {
        if (question == null) return "general";

        String lowercase = question.toLowerCase();
        for (String word : QUESTION_WORDS) {
            lowercase = lowercase.replace(word, "");
        }

        String[] sentences = lowercase.split("[.,]");
        String firstSentence = sentences.length > 0 ? sentences[0] : "";
        return firstSentence.trim().isEmpty() ? "general" : firstSentence.trim();
    }
}